python/cudf/cudf/core/df_protocol.py

```diff
@@ -0,0 +1,729 @@
+    _np_dtypes = {0: _ints, 1: _uints, 2: _floats, 20: {8: bool}}
+    column_dtype = _np_dtypes[kind][bitwidth]
+
+    # No DLPack yet, so need to construct a new ndarray from the data pointer
+    # and size in the buffer plus the dtype on the column
+    ctypes_type = np.ctypeslib.as_ctypes_type(column_dtype)
+    data_pointer = ctypes.cast(_buffer.ptr, ctypes.POINTER(ctypes_type))
+
+    # NOTE: `x` does not own its memory, so the caller of this function must
+    #       either make a copy or hold on to a reference of the column or
+    #       buffer! (not done yet, this is pretty awful ...)
+    x = np.ctypeslib.as_array(data_pointer,
+                              shape=(_buffer.bufsize // (bitwidth//8),))
+    return cp.asarray(x, dtype=column_dtype)
+
+
+def convert_categorical_column(col : ColumnObject, allow_copy:bool=False) :
+    """
+    Convert a categorical column to a Series instance
+    """
+    ordered, is_dict, mapping = col.describe_categorical
+    if not is_dict:
+        raise NotImplementedError('Non-dictionary categoricals not supported yet')
+
+    # If you want to cheat for testing (can't use `_col` in real-world code):
+    #    categories = col._col.values.categories.values
+    #    codes = col._col.values.codes
+    categories = cp.asarray(list(mapping.values()))
+    codes_buffer, codes_dtype = col.get_buffers()['data']
+    codes = buffer_to_cupy_ndarray(codes_buffer, codes_dtype, 
+                                   allow_copy=allow_copy)
+    values = categories[codes]
+
+    # Seems like cudf can only construct with non-null values, so need to
+    # null out the nulls later
+    cat = cudf.CategoricalIndex(values, categories=categories, ordered=ordered)
+    return set_missing_values(col, cat), codes_buffer
+
+
+def __dataframe__(self, nan_as_null : bool = False,
+                  allow_copy : bool = True) -> dict:
+    """
+    The public method to attach to cudf.DataFrame.
+
+    We'll attach it via monkey-patching here for demo purposes. If Pandas adopts
+    the protocol, this will be a regular method on pandas.DataFrame.
+
+    ``nan_as_null`` is a keyword intended for the consumer to tell the
+    producer to overwrite null values in the data with ``NaN`` (or ``NaT``).
+    This currently has no effect; once support for nullable extension
+    dtypes is added, this value should be propagated to columns.
+
+    ``allow_copy`` is a keyword that defines whether or not the library is
+    allowed to make a copy of the data. For example, copying data would be
+    necessary if a library supports strided buffers, given that this protocol
+    specifies contiguous buffers.
+    Currently, if the flag is set to ``False`` and a copy is needed, a
+    ``RuntimeError`` will be raised.
+    """
+    return _CuDFDataFrame(
+        self, nan_as_null=nan_as_null, allow_copy=allow_copy)
+
+
+# Implementation of interchange protocol
+# --------------------------------------
+
+class _CuDFBuffer:
```

Comment:
I observe that `_CuDFBuffer` maps more closely to [`cudf.core.buffer.Buffer`](https://docs.rapids.ai/api/cudf/stable/internals.html#buffer). Similarly, `_CuDFColumn` below maps more closely to [`cudf.core.column.Column`](https://docs.rapids.ai/api/cudf/stable/internals.html#column).

Could these classes be implemented on top of those types? Alternatively, could those types just be used here instead?

Reply:
Thank you for your good remarks and interesting questions. 
First of all, the implementation is up to the supporting library (here `cudf`). Thus, there is no inconvenience to derive _Buffer/_Column from their cudf counterparts as long as we only expose the [protocol methods](https://github.com/data-apis/dataframe-api/blob/main/protocol/dataframe_protocol.py). Secondly, this implementation is heavily inspired from the [pandas' one](https://github.com/data-apis/dataframe-api/blob/main/protocol/pandas_implementation.py). That said,

- Concerning `_CuDFColumn`, note that it has an attribute `_col` which is a ` cudf.core.column.Column` object. We used it to expose only the [protocol methods](https://github.com/data-apis/dataframe-api/blob/main/protocol/dataframe_protocol.py).
- For `_CuDFBuffer`, what mainly refrain me from using/deriving from `cudf.core.buffer.Buffer` was the absence of the [`to_dlpack` method](https://github.com/data-apis/dataframe-api/blob/5cbbfc1fcc6458e367ba55b28196e82c87014311/protocol/dataframe_protocol.py#L30)  that is available on the `cupy` array object which an attribute of `_CuDFBuffer`. 

I think we indirectly derive from `cudf` object to implement the protocol using composition instead of inheritance.
I hope this answer your questions.

Reply:
Thanks! Agree, that composition v/s inheritance is the better approach here.

> Concerning _CuDFColumn, note that it has an attribute _col which is a cudf.core.column.Column object.

In the current implementation, it looks like `_col` would end up being a `Series` object, as opposed to a `Column` object. For example, the `iloc` operation [here](https://github.com/rapidsai/cudf/blob/afdfeedeb8ea76af6137b0b492d13833e698155c/python/cudf/cudf/core/df_protocol.py#L702) yields a `Series`, which is passed to the `_CuDFColumn` constructor.

> For _CuDFBuffer, what mainly refrain me from using/deriving from cudf.core.buffer.Buffer was the absence of the to_dlpack method

Would something like the following work for implementing `__dlpack__`?

```python
>>> buf = cudf.core.buffer.Buffer(np.array([1, 2, 3], dtype="uint8"))
>>> cp.asarray(buf).toDlpack()
<capsule object "dltensor" at 0x7febec690fc0>
```

Ideally, we could avoid using a `cp.array()` underneath a `CuDFBuffer` here, since the latter represents an _untyped_ memory allocation on the GPU (which is precisely what `cudf.core.buffer.Buffer` is).

Reply:
Apologies, I made a mistake. In fact the `self._col` attribute is always a `cudf.core.series.Series` object! We have this line in the constructor: 
```
if not isinstance(column, cudf.Series):
    raise NotImplementedError("Columns of type {} not handled "
                                                "yet".format(type(column)))
```
 But in this case both series and column have necessary methods to implement the protocol.  Can we just use the `_column` attribute of cudf series which returns a cudf column i.e. `self._df.iloc[:, i]._column`  or `cudf.core.column.as_column(self._df.iloc[:, i])`?

Concerning the buffer, your suggestion would definitely work! Also, it can reduce some overheads as of now, we first get the array attached to the column in order to create the `_CuDFBuffer`: 
`_CuDFBuffer(cp.asarray(self._col.fillna(invalid).to_gpu_array(), dtype=cp.uint8), allow_copy=self._allow_copy)`
So, we can use the column buffer directly `self._col._data` and use it as an attribute instead of the array and that will improve the performance I guess: 
`_CuDFBuffer(self._col._data, allow_copy=self._allow_copy)`
However, I see some issues with that approach:
- when there is a missing value, `col._data` return `None` which will break our `_CuDFBuffer` object.
- Even if I fill in missing values with valid ones before accessing the buffer with `_data`, I get the same result ! Is that  expected or a silent bug ??
- Some other times it just does return `None`:
```
>>> df = cudf.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
>>> cudf.core.column._as_column(df['a'])._data
>>>
```
Is there a safe way to access the data buffer associated to a cudf Column or Series ? Or shall we take the array as for now and then construct the buffer out of it ?

Reply:
Yes -- you should access the public `.data` attribute instead:

```python

In [2]: df = cudf.DataFrame({'a': [1,2,3], 'b': [4,5,6]})

In [3]: cudf.core.column.as_column(df['a'])._data

In [4]: cudf.core.column.as_column(df['a']).data
Out[4]: <cudf.core.buffer.Buffer at 0x7efc101f40d0>
```